AzureMonitor: Application Insights Traces

aangelisc opened the pull request:
On completion, this PR will add support for Application Insights Traces. Users will be able to retrieve traces from relevant resources and view them using the traces visualisation. 

For now this PR will act as the base branch with other Traces PR's being merged here.

PR's included:

#65442
#65566
#66076
#65855
#66303
#66474
#67292 

Closes #63131, #63117, #63118, #63119, #63121, #63124, #63127, #63129

aangelisc committed:
Build out barebones Traces editor

- Add Traces query type and operation ID prop to query type
- Add necessary header types
- Update resource picker to appropriately work with traces query type
- Build out TracesQueryEditor component
- Include logic to retrieve operationId's for AI Workspaces
- Add backend route mapping
- Update macro to use timestamp as default time field for traces

bossinc commented on public/app/plugins/datasource/azuremonitor/components/TracesQueryEditor/setQueryValue.ts:
```diff
@@ -0,0 +1,35 @@
+import { AzureMonitorQuery, ResultFormat } from '../../types';
+
+const buildTracesQuery = (operationId: string): string =>
+  `set truncationmaxrecords=10000;
+    set truncationmaxsize=67108864;
+    union isfuzzy=true traces, customEvents, pageViews, requests, dependencies, exceptions, availabilityResults
+    | where $__timeFilter()
+    | where (operation_Id != '' and operation_Id == '${operationId}') or (customDimensions.ai_legacyRootId != '' and customDimensions.ai_legacyRootId == '${operationId}')
+    | extend duration = iff(isnull(duration), toreal(0), duration)
+    | extend spanID = iff(itemType == "pageView" or isempty(id), tostring(new_guid()), id)
+    | extend serviceName = iff(isempty(name), column_ifexists("problemId", ""), name)
+    | project operation_Id, operation_ParentId, itemType, spanID, duration, timestamp, serviceName, customDimensions
+    | project-rename traceID = operation_Id, parentSpanID = operation_ParentId, startTime = timestamp, serviceTags = customDimensions`;
```
Why is this being set here and not by the user? I think I'm not understanding how this all works

aangelisc replied:
This query is a generic query for retrieving a specific trace based on `operationId`. A user wouldn't need to really modify it to retrieve that trace. Also this is just the barebones editor at the moment so the finished product might look different 😉

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
AzureMonitor: Traces - Response parsing (#65442)

* Update FormatAsField component

- Add trace ResultFormat type
- Generalise FormatAsField component
- Add component to TracesQueryEditor
- Remove duplicate code in setQueryValue

* Add custom filter function to improve performance

* Add basic conversion for logs to trace

- Add serviceTags converter
- Pass through required parameters (queryType and resultFormat)
- Appropriately set visualisation

* Update parsing to also fill trace tags

- Add constant values for each table schema (include legacy mapping for now if needed)
- Add constant for list of table tags
- Set the foundation for dynamic query building
- Update query to build tags value
- Appropriately set operationName
- Update tagsConverter to filter empty values

* Fix lint and test issues

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

# Conflicts:
#	pkg/tsdb/azuremonitor/loganalytics/azure-log-analytics-datasource.go

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

bossinc commented on pkg/tsdb/azuremonitor/loganalytics/azure-response-table-frame.go:
```diff
@@ -112,6 +116,55 @@ var converterMap = map[string]data.FieldConverter{
+			return nil, nil
+		}
+
+		m := map[string]interface{}{}
```
```suggestion
		m := map[string]any{}
```
You can use `any` instead of empty interfaces

bossinc reviewed (commented):
This is great!

aangelisc committed:
AzureMonitor: Traces - Data links (#65566)

* Add portal link for traces

- Pull out necessary values (itemId and itemType)
- Appropriately construct
- Fix ordering

* Set default format as value

- Also set default visualisation

* Fix event schema

* Set default formatAsField value

* Include logs link on traces results

- Adapt config links to allow custom title to be set

* Correctly set operationId for query

* Update backend types

- Include OperationID in query
- Pass forward datasource name and UID

* Ensure setTime doesn't consistently get called if operationID is defined

* Add explore link

- Update util functions to allow setting custom datalinks

* Fix tests

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
AzureMonitor: Traces - Query and Editor updates (#66076)

* Add initial query

- Will query the resource as soon as a resource has been selected
- Updates the data links for the query without operationId
- Remove initial operationId query and timeRange dependency
- Update query building

* Add entirely separate traces query property

- Update shared types (also including future types for Azure traces)
- Update backend log analytics datasource to accept both azureLogAnalytics and azureTraces queries
- Update backend specific types
- Update frontend datasource for new properties
- Update mock query

* Update FormatAsField to be entirely generic

* Update query building to be done in backend

- Add required mappings in backend
- Update frontend querying

* Fix query and explore data link

* Add trace type selection

* Better method for setting explore link

* Fix operationId updating

* Run go mod tidy

* Unnecessary changes

* Fix tests

aangelisc committed:
AzureMonitor: Traces - Add correlation API support (#65855)

Add correlation API support

- Add necessary types
- Add correlation API request when conditions are met
- Update query

aangelisc committed:
Fix property from merge

aangelisc committed:
AzureMonitor: Traces - Filtering (#66303)

* Add initial query

- Will query the resource as soon as a resource has been selected
- Updates the data links for the query without operationId
- Remove initial operationId query and timeRange dependency
- Update query building

* Add entirely separate traces query property

- Update shared types (also including future types for Azure traces)
- Update backend log analytics datasource to accept both azureLogAnalytics and azureTraces queries
- Update backend specific types
- Update frontend datasource for new properties
- Update mock query

* Update FormatAsField to be entirely generic

* Update query building to be done in backend

- Add required mappings in backend
- Update frontend querying

* Fix query and explore data link

* Add trace type selection

* Better method for setting explore link

* Fix operationId updating

* Run go mod tidy

* Unnecessary changes

* Fix tests

* Start building out Filters component

- Configure component to query for Filter property values when a filter property is set
- Add setFilters function
- Add typing to tablesSchema
- Use component in TracesQueryEditor

* Update Filters

- Asynchronously pull property options
- Setup list of Filter components

* Update filters component

- Remove unused imports
- Have local filters state and query filters
- Correctly set filters values
- Don't update query every time a filter property changes (not performant)

* Update properties query

- Use current timeRange
- Get count to provide informative labels

* Reset map when time changes

* Add operation selection

* Reset filters when property changes

* Appropriate label name for empty values

* Add filtering to query

* Update filter components

- Fix rendering issue
- Correctly compare and update timeRange
- Split out files for simplicity

* Add checkbox option to multiselect

- Add custom option component
- Correctly call onChange
- Add variableOptionGroup for template variable selection

* Fix adding template vars

* Improve labels and refresh labels on query prop changes

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
Merge branch 'main' into azure-monitor-traces

aangelisc committed:
AzureMonitor: Traces - Testing (#66474)

* Select ds for template variable interpolation

* Update az logs ds tests

- Add templateVariables test
- Add filter test
- Update mock
- Remove anys

* Update QueryEditor test

- Update mocks with timeSrv for log analytics datasource
- Fix query mock
- Use appropriate and consistent selectors

* Add TracesQueryEditor test

- Update resourcePickerRows mock to include app insights resources
- Remove comments and extra new line

* Add FormatAsField test

- Remove unneeded condition

* Update resourcePicker utils test

* Don't hide selected options in filters

* Fix multi-selection on filters

* Add TraceTypeField test

- Add test file
- Update selectors (remove copy/paste mistake)
- Update placeholder text for select and add label

* Add basic filters test

* Begin filters test

* Update filters test

* Add final tests and simplify/generalise addFilter helper

* Minor update to datasource test

* Update macros test

* Update selectors in tests

* Add response-table-frame tests

* Add datasource tests

- Use sorting where JSON models are inconsistent
- Update filters clause
- Dedupe tags
- Correct operationId conditions

* Don't set a default value for blurInputOnSelect

* Simplify datasource test

* Update to use CheckGoldenJSON utils

- Update with generated frame files
- Remove redundant expected frame code
- Update all usages

* Fix lint

aangelisc committed:
AzureMonitor: Traces feedback (#67292)

* Filter traces if the visualisation is set to trace

- Update build query logic
- Added additional test cases
- Return an error if the traces type is set by itself with the trace visualisation
- Add descriptions to event types
- Update tests

* Fix bug for error displaying traces

* Update mappings and add error field

- Update tests
- Remove unnecessary comments

* Switch location of Operation ID field

* Re-order fields

* Update link title

* Update label for event type selection

* Update correct link title

* Update logs datalink to link to Azure Logs in explore

* Fix lint

asimpson reviewed (approved):
Nicely done dude :trophy:
